zcash_client_sqlite/src/wallet.rs

```diff
@@ -870,86 +1074,114 @@ fn subtree_scan_progress<P: consensus::Parameters>(
+    if fully_scanned_height == Some(chain_tip_height) {
         // Compute the total blocks scanned since the wallet birthday on either side of
         // the recover-until height.
-        let recover = recover_until_height
-            .map(|end_height| {
-                stmt_scanned_count_between.query_row(
-                    named_params! {
-                        ":start_height": u32::from(birthday_height),
-                        ":end_height": u32::from(end_height),
-                    },
-                    |row| {
-                        let recovered = row.get::<_, Option<u64>>(0)?;
-                        Ok(recovered.map(|n| Ratio::new(n, n)))
-                    },
-                )
-            })
-            .transpose()?
-            // If none of the wallet's accounts have a recover-until height, then we can't
-            // (yet) distinguish general scanning from recovery, so treat the wallet as
-            // fully recovered.
-            .unwrap_or_else(|| Some(Ratio::new(1, 1)));
+        let recover = match recover_until_height {
+            Some(end_height) => stmt_scanned_count_until.query_row(
+                named_params! {
+                    ":start_height": u32::from(birthday_height),
+                    ":end_height": u32::from(end_height),
+                },
+                |row| {
+                    let recovered = row.get::<_, Option<u64>>(0)?;
+                    Ok(recovered.map(|n| Ratio::new(n, n)))
+                },
+            )?,
+            None => {
+                // If none of the wallet's accounts have a recover-until height, then there
+                // is no recovery phase for the wallet, and therefore the denominator in the
+                // resulting ratio (the number of notes in the recovery range) is zero.
+                Some(Ratio::new(0, 0))
+            }
+        };
+
         let scan = stmt_scanned_count_from.query_row(
             named_params! {
                 ":start_height": u32::from(
-                    recover_until_height.map(|h| h + 1)
-                        .unwrap_or(birthday_height)
+                    recover_until_height.unwrap_or(birthday_height)
                 ),
             },
             |row| {
                 let scanned = row.get::<_, Option<u64>>(0)?;
                 Ok(scanned.map(|n| Ratio::new(n, n)))
             },
         )?;
-        Ok(Progress { scan, recover })
+
+        Ok(scan.map(|scan| Progress::new(scan, recover)))
     } else {
+        // In case we didn't have information about the tree size at the recover-until
+        // height, get the tree size from a nearby subtree. It's fine for this to be
+        // approximate; it just shifts the boundary between scan and recover progress.
+        let mut get_tree_size_near = |as_of: BlockHeight| {
+            let size_from_blocks = stmt_start_tree_size
+                .query_row(named_params![":start_height": u32::from(as_of)], |row| {
```

Comment:
Good catch on this! (`as_of` was incorrectly `birthday_height` before the most recent push.)